python/tank/context.py

```diff
@@ -668,6 +668,83 @@ def create_empty(tk):
     """
     return Context(tk)
 
+def from_entities(tk, project, entity, step, task, user):
+    """
+    Construct a new context from the specified entities.
+
+    Note: It is the responsibility of the calling code to ensure that both the entities and the
+    relationships between them are valid.  
+    
+    This only does a Shotgun query if the context_additional_entities hook specifies additional
+    fields on a task to look for additional entities on!
+
+    :param tk:      An sgtk API instance
+    :param project: The project entity dictionary to use when constructing the context
+    :param entity:  The entity entity dictionary to use when constructing the context
+    :param step:    The step entity dictionary to use when constructing the context
+    :param task:    The task entity dictionary to use when constructing the context
+    :param user:    The user entity dictionary to use when constructing the context
+    :returns:       A Context constructed from the specified entities.
+    """
+    # prep our return data structure
+    context = {
+        "tk": tk,
+        "project": None,
+        "entity": None,
+        "step": None,
+        "user": None,
+        "task": None,
+        "additional_entities": []
+    }    
+
+    if project and "type" in project and "id" in project:
+        context["project"] = {
+            "type":project["type"], 
+            "id":project["id"], 
+            "name":project.get("name")
+        }
+
+    if entity and "type" in entity and "id" in entity and entity["type"] != "Project":
+        name_field = {"HumanUser":"name", "Task":"content"}.get(entity["type"], "code")
+        context["entity"] = {
+            "type":entity["type"], 
+            "id":entity["id"], 
+            # name can either be the name field or 'name' if the entity dict has come from another context
+            "name":(entity.get(name_field) or project.get("name"))
+        }
+
+    if step and "type" in step and "id" in step:
+        context["step"] = {
+            "type":step["type"], 
+            "id":step["id"],
+            # name can either be 'code' or 'name' if the entity dict has come from another context
+            "name":(step.get("code") or step.get("name"))
+        }
+        
+    if user and "type" in user and "id" in user:
+        context["user"] = {
+            "type":user["type"], 
+            "id":user["id"], 
+            "name":user.get("name")
+        }
+
+    if task and "type" in task and "id" in task:
+        # first check to see if there are any additional entities specified:
+        additional_fields = tk.execute_core_hook("context_additional_entities").get("entity_fields_on_task", [])
```

Comment:
This is really just to avoid executing the hook twice when this is a pretty slow piece of code anyway!